File: 503-next-greater-element-ii/next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n=nums.length;
        int[] ans=new int[n];
        for(int i=0;i<n;i++) ans[i]=-1;
        Stack<Integer> s=new Stack<>();
        s.push(0);
        int l=2*n-1;

        for(int i=1;i<=l;i++){
            System.out.println(i+" and "+i%n);

            int c=s.peek();
            
            if(nums[c]<nums[i%n]){
                while(s.size()!=0 && nums[s.peek()]<nums[i%n]){
                    int k=s.pop();
                    ans[k]=nums[i%n];
                }
                s.push(i%n);
            }
            else
            s.push(i%n);
        }
        return ans;
    }
}

File: 503-next-greater-element-ii/next-greater-element-ii.py
class Solution(object):
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        
        stack = []
        n = len(nums)
        ans = [-1 for i in range(n)]

        for i in range(2* n):
            ind = i % n

            while stack and nums[stack[-1]] < nums[ind]:
                j = stack.pop(-1)
                if ans[j] == -1: ans[j] = nums[ind]
            stack.append(ind)

        return ans